confignode/src/main/java/org/apache/iotdb/confignode/procedure/env/DataNodeRemoveHandler.java

```diff
@@ -135,29 +136,99 @@ public TDataNodeLocation findDestDataNode(TConsensusGroupId regionId) {
       return status;
     }
 
-    // TODO if region replica is 1, the new leader is null, it also need to migrate
-    Optional<TDataNodeLocation> newLeaderNode =
-        regionReplicaNodes.stream().filter(e -> !e.equals(originalDataNode)).findAny();
+    TMigrateRegionReq migrateRegionReq =
+        new TMigrateRegionReq(regionId, destDataNode, destDataNode);
+    migrateRegionReq.setNewLeaderNode(newLeaderNode.get());
+
+    // send to new leader
+    status =
+        SyncDataNodeClientPool.getInstance()
+            .sendSyncRequestToDataNodeWithRetry(
+                newLeaderNode.get().getInternalEndPoint(),
+                migrateRegionReq,
+                DataNodeRequestType.ADD_REGION_PEER);
+    LOGGER.info(
+        "Send region {} migrate action to {}, wait it finished",
+        regionId,
+        destDataNode.getInternalEndPoint());
+    return status;
+  }
+
+  /**
+   * Send to DataNode, remove region
+   *
+   * @param originalDataNode old location data node
+   * @param destDataNode dest data node
+   * @param regionId region id
+   * @return migrate status
+   */
+  public TSStatus removeRegionPeer(
+      TDataNodeLocation originalDataNode,
+      TDataNodeLocation destDataNode,
+      TConsensusGroupId regionId) {
+    TSStatus status;
+    Optional<TDataNodeLocation> newLeaderNode = findNewLeader(regionId, originalDataNode);
+    if (!newLeaderNode.isPresent()) {
+      LOGGER.warn(
+          "No other Node to change region leader, check by show regions, region: {}", regionId);
+      status = new TSStatus(TSStatusCode.MIGRATE_REGION_ERROR.getStatusCode());
+      status.setMessage("No other Node to change region leader, check by show regions");
+      return status;
+    }
+
+    TMigrateRegionReq migrateRegionReq =
+        new TMigrateRegionReq(regionId, originalDataNode, destDataNode);
+    migrateRegionReq.setNewLeaderNode(newLeaderNode.get());
+
+    // send to new leader
+    status =
+        SyncDataNodeClientPool.getInstance()
+            .sendSyncRequestToDataNodeWithRetry(
+                newLeaderNode.get().getInternalEndPoint(),
+                migrateRegionReq,
+                DataNodeRequestType.REMOVE_REGION_PEER);
+    LOGGER.info(
+        "Send region {} remove peer to {}, wait it finished",
+        regionId,
+        originalDataNode.getInternalEndPoint());
+    return status;
+  }
+
+  /**
+   * Send to DataNode, remove region consensus group from originalDataNode node
+   *
+   * @param originalDataNode old location data node
+   * @param destDataNode dest data node
+   * @param regionId region id
+   * @return migrate status
+   */
+  public TSStatus removeRegionConsensusGroup(
+      TDataNodeLocation originalDataNode,
+      TDataNodeLocation destDataNode,
+      TConsensusGroupId regionId) {
+    TSStatus status;
+    Optional<TDataNodeLocation> newLeaderNode = findNewLeader(regionId, originalDataNode);
```

Comment:
Why should we findNewLeader here since we've executed changeLeader at an earlier step in RemoveDataNodeProcedure?

Reply:
rename find other node.